src/environmentd/src/http/sql.rs

```diff
@@ -47,6 +47,61 @@ use tungstenite::protocol::frame::coding::CloseCode;
 
 use crate::http::{init_ws, AuthedClient, AuthedUser, WsState, MAX_REQUEST_SIZE};
 
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error(transparent)]
+    Adapter(AdapterError),
+    #[error(transparent)]
+    Json(#[from] serde_json::Error),
+    #[error(transparent)]
+    Axum(#[from] axum::Error),
+    #[error("SUBSCRIBE only supported over websocket")]
+    SubscribeOnlyOverWs,
+    #[error("current transaction is aborted, commands ignored until end of transaction block")]
+    AbortedTransaction,
+    #[error("unsupported via this API: {0}")]
+    Unsupported(String),
+    #[error("{0}")]
+    Unstructured(anyhow::Error),
+}
+
+impl Error {
+    pub fn detail(&self) -> Option<String> {
+        match self {
+            Error::Adapter(err) => err.detail(),
+            _ => None,
+        }
+    }
+
+    pub fn hint(&self) -> Option<String> {
+        match self {
+            Error::Adapter(err) => err.hint(),
+            _ => None,
+        }
+    }
+
+    pub fn position(&self) -> Option<usize> {
+        match self {
+            Error::Adapter(err) => err.position(),
+            _ => None,
+        }
+    }
+
+    pub fn code(&self) -> SqlState {
+        match self {
+            Error::Adapter(err) => err.code(),
+            Error::AbortedTransaction => SqlState::IN_FAILED_SQL_TRANSACTION,
+            _ => SqlState::INTERNAL_ERROR,
+        }
+    }
+}
+
+impl From<AdapterError> for Error {
+    fn from(value: AdapterError) -> Self {
+        Self::Adapter(value)
+    }
+}
```

Comment:
Since we dropped the `Box<AdapterError>` I this can go back to using `#[from]` in `enum Error`, unless this was causing the infinite recursion!

Reply:
Good call, fixed.